jlm/mlir/backend/mlirgen.cpp

```diff
@@ -0,0 +1,269 @@
+  auto lambdaResult = builder->create<mlir::rvsdg::LambdaResult>(builder->getUnknownLoc(), *regionResults);
+  lambdaBlock->push_back(lambdaResult);
+
+  return lambda;
+}
+
+mlir::Type
+MLIRGen::convertType(const rvsdg::type *type)
+{
+  if(auto bt = dynamic_cast<const rvsdg::bittype *>(type)){
+    return builder->getIntegerType(bt->nbits());
+  } else if (auto loopstateType = dynamic_cast<const llvm::loopstatetype*>(type)){
+    return builder->getType<::mlir::rvsdg::LoopStateEdgeType>();
+  } else if (auto iostateType = dynamic_cast<const llvm::iostatetype*>(type)){
+    return builder->getType<::mlir::rvsdg::IOStateEdgeType>();
+  } else if (auto memstateType = dynamic_cast<const llvm::MemoryStateType*>(type)){
+    return builder->getType<::mlir::rvsdg::MemStateEdgeType>();
+  /*
+  } else if (auto varargType =dynamic_cast<const jlm::varargtype*>(type)) {
+    s << "!jlm.varargList";
+  } else if (auto pointerType = dynamic_cast<const jlm::PointerType*>(type)){
+    s << print_pointer_type(pointer_type);
+  } else if (auto arrayType = dynamic_cast<const jlm::arraytype*>(type)){
+    s << print_array_type(array_type);
+  } else if (auto structType = dynamic_cast<const jlm::structtype*>(type)){
+    s << print_struct_type(struct_type);
+  } else if (auto controlType = dynamic_cast<const jive::ctltype*>(type)){
+    s << "!rvsdg.ctrl<" << control_type->nalternatives() << ">";
+  */
+  } else{
+    throw util::error("Type conversion not implemented: " +
+		     type->debug_string());
+  }
+}
+
+// TODO
+// This function might not be needed to generate MLIR
+// But kept for now in case it is needed
+mlir::Value
+MLIRGen::convertArgument(const rvsdg::argument *argument)
+{
+  if(auto bt = dynamic_cast<const rvsdg::bittype *>(&argument->type())){
+    //    s << "i" << bt->nbits();
+  /*
+  } else if (auto loopstateType = dynamic_cast<const jlm::loopstatetype*>(type)){
+    s << "!rvsdg.loopState";
+  } else if (auto iostateType = dynamic_cast<const jlm::iostatetype *>(&argument->type())){
+//    return  "!rvsdg.ioState";
+  } else if (auto memstateType = dynamic_cast<const jlm::MemoryStateType*>(type)){
+    */
+//   return builder->create(builder->getUnknownLoc(), convertType(&argument->type()));
+    /*
+  } else if (auto varargType =dynamic_cast<const jlm::varargtype*>(type)) {
+    s << "!jlm.varargList";
+  } else if (auto pointerType = dynamic_cast<const jlm::PointerType*>(type)){
+    s << print_pointer_type(pointer_type);
+  } else if (auto arrayType = dynamic_cast<const jlm::arraytype*>(type)){
+    s << print_array_type(array_type);
+  } else if (auto structType = dynamic_cast<const jlm::structtype*>(type)){
+    s << print_struct_type(struct_type);
+  } else if (auto controlType = dynamic_cast<const jive::ctltype*>(type)){
+    s << "!rvsdg.ctrl<" << control_type->nalternatives() << ">";
+  */
+  } else{
+    throw util::error("Argument conversion not implemented: " +
+		     argument->type().debug_string());
+  }
+  return NULL;
```

Comment:
The return value is `mlir::Value`, but you return `NULL` here (which should `nullptr`), which is normally a pointer. Does this even compile?

Reply:
I guess the only thing that safes you here is the exception in the else case.